File: src/agent/agent-sdk/python/bkci_agent_sdk/docker_build.py
"""Docker/Podman worker build orchestration."""

from __future__ import annotations

import asyncio
from dataclasses import dataclass, field
from pathlib import Path
import random
import string
import sys
from typing import Callable, Mapping

from .docker_cli import DockerRunner
from .types import DockerOptions, ImagePullPolicy, ThirdPartyBuildInfo

CONTAINER_WORKER_JAR = "/data/worker-agent.jar"
ENTRY_POINT_CMD = "/data/init.sh"
TARGET_JRE_DIR = "/usr/local/jre"
TARGET_JRE8_DIR = "/usr/local/jre8"
DOCKER_LOG_DIR = "/data/devops/logs"
DOCKER_DATA_DIR = "/data/devops/workspace"
DOCKER_NO_MOUNT = "__NO_MOUNT__"

_LOCAL_DOCKER_WORKSPACE = "docker_workspace"
_LOCAL_DOCKER_BUILD_TEMP = "docker_build_tmp"
_LONG_LOG_TAG = "toolong"


@dataclass(slots=True)
class DockerBuildOptions:
    build_info: ThirdPartyBuildInfo
    work_dir: str | Path
    worker_jar_path: str | Path
    docker_init_script_path: str | Path
    gateway: str
    project_id: str
    jdk17_dir_path: str | Path | None = None
    jdk8_dir_path: str | Path | None = None
    extra_env: Mapping[str, str] = field(default_factory=dict)
    platform: str | None = None
    post_log: Callable[[str], None] | None = None
    rand_suffix: Callable[[], str] | None = None


@dataclass(slots=True)
class DockerBuildResult:
    success: bool
    message: str


def _ensure_gateway(gateway: str) -> str:
    return gateway if gateway.startswith(("http://", "https://")) else "http://" + gateway


def _parse_api_docker_options(options: DockerOptions) -> list[str]:
    args: list[str] = []
    for volume in options.get("volumes", []) or []:
        if volume.strip():
            args.extend(["--volume", volume.strip()])
    for mount in options.get("mounts", []) or []:
        if mount.strip():
            args.extend(["--mount", mount.strip()])
    gpus = str(options.get("gpus", "") or "").strip()
    if gpus:
        args.extend(["--gpus", gpus])
    if options.get("privileged"):
        args.append("--privileged")
    for network in options.get("network", []) or []:
        if network.strip():
            args.extend(["--network", network.strip()])
    user = str(options.get("user", "") or "").strip()
    if user:
        args.extend(["--user", user])
    return args


def _build_user_docker_args(options: DockerOptions) -> list[str]:
    args = _parse_api_docker_options(options)
    index = 0
    options_with_values = {"--volume", "--mount", "--network", "--user", "--gpus"}
    while index < len(args):
        if args[index] in options_with_values:
            if index + 1 >= len(args) or not args[index + 1].strip():
                raise ValueError(f"docker option {args[index]} requires a non-empty value")
            index += 2
        else:
            index += 1
    return args


def _has_custom_network(options: DockerOptions) -> bool:
    return len(options.get("network", []) or []) > 0


def _need_local_image_inspect(is_latest: bool, policy: str) -> bool:
    if policy == ImagePullPolicy.ALWAYS.value:
        return False
    if policy == ImagePullPolicy.IF_NOT_PRESENT.value:
        return True
    return not is_latest


def _if_pull_image(local_exists: bool, is_latest: bool, policy: str) -> bool:
    if policy == ImagePullPolicy.ALWAYS.value:
        return True
    if policy == ImagePullPolicy.IF_NOT_PRESENT.value:
        return not local_exists
    return True if is_latest else not local_exists


def _has_jdk17_dir(path: str | Path | None) -> bool:
    return bool(path) and Path(path).is_dir()


def _container_env(options: DockerBuildOptions) -> list[str]:
    docker = options.build_info.get("dockerBuildInfo") or {}
    values = [
        f"devops_project_id={options.project_id}",
        f"devops_agent_id={docker.get('agentId', '')}",
        f"devops_agent_secret_key={docker.get('secretKey', '')}",
        f"devops_gateway={_ensure_gateway(options.gateway)}",
        "agent_build_env=DOCKER",
    ]
    if _has_jdk17_dir(options.jdk17_dir_path) and (options.platform or sys.platform) == "linux":
        values.extend(
            [
                f"DEVOPS_AGENT_JDK_8_PATH={TARGET_JRE8_DIR}/bin/java",
                f"DEVOPS_AGENT_JDK_17_PATH={TARGET_JRE_DIR}/bin/java",
            ]
        )
    values.extend(f"{key}={value}" for key, value in options.extra_env.items())
    return values


async def _container_mount_args(options: DockerBuildOptions) -> list[str]:
    build = options.build_info
    args: list[str] = []
    if _has_jdk17_dir(options.jdk17_dir_path):
        args.extend(
            [
                "--mount",
                f"type=bind,source={options.jdk17_dir_path},target={TARGET_JRE_DIR},readonly",
                "--mount",
                f"type=bind,source={options.jdk8_dir_path or ''},target={TARGET_JRE8_DIR},readonly",
            ]
        )
    elif options.jdk8_dir_path:
        args.extend(
            [
                "--mount",
                f"type=bind,source={options.jdk8_dir_path},target={TARGET_JRE_DIR},readonly",
            ]
        )
    args.extend(
        [
            "--mount",
            f"type=bind,source={options.worker_jar_path},target={CONTAINER_WORKER_JAR},readonly",
            "--mount",
            f"type=bind,source={options.docker_init_script_path},target={ENTRY_POINT_CMD},readonly",
        ]
    )

    work_dir = Path(options.work_dir)
    data_dir = work_dir / _LOCAL_DOCKER_WORKSPACE / "data" / str(
        build.get("pipelineId", "")
    ) / str(build.get("vmSeqId", ""))
    if build.get("workspace"):
        data_dir = Path(str(build["workspace"]))
    if str(data_dir) != DOCKER_NO_MOUNT:
        data_dir.mkdir(parents=True, exist_ok=True)
        args.extend(["--mount", f"type=bind,source={data_dir},target={DOCKER_DATA_DIR}"])

    logs_dir = (
        work_dir
        / _LOCAL_DOCKER_WORKSPACE
        / "logs"
        / str(build.get("buildId", ""))
        / str(build.get("vmSeqId", ""))
    )
    logs_dir.mkdir(parents=True, exist_ok=True)
    args.extend(["--mount", f"type=bind,source={logs_dir},target={DOCKER_LOG_DIR}"])
    return args


async def build_docker_create_args(
    container_name: str, image: str, options: DockerBuildOptions
) -> list[str]:
    docker = options.build_info.get("dockerBuildInfo") or {}
    docker_options: DockerOptions = docker.get("options") or {}
    mount_args = await _container_mount_args(options)
    user_args = _build_user_docker_args(docker_options)
    args = ["--name", container_name, *user_args]
    if not _has_custom_network(docker_options):
        args.extend(["--network", "bridge"])
    for value in _container_env(options):
        args.extend(["-e", value])
    args.extend(mount_args)
    args.extend(["--entrypoint", "/bin/sh", image, "-c", ENTRY_POINT_CMD])
    return args


async def run_docker_build(
    runner: DockerRunner, options: DockerBuildOptions
) -> DockerBuildResult:
    build = options.build_info
    docker = build.get("dockerBuildInfo")
    post_log = options.post_log or (lambda message: None)
    if not docker:
        return DockerBuildResult(False, "dockerBuildInfo is missing")
    credential = docker.get("credential") or {}
    if credential.get("errMsg"):
        return DockerBuildResult(False, f"get docker cred error: {credential['errMsg']}")

    image_name = str(docker.get("image", "") or "").strip()
    image = image_name.removeprefix("http://").removeprefix("https://")
    image_parts = image.split(":")
    is_latest = len(image_parts) == 1 or (
        len(image_parts) == 2 and image_parts[1] == "latest"
    )
    raw_policy = docker.get("imagePullPolicy", "")
    policy = raw_policy.value if isinstance(raw_policy, ImagePullPolicy) else str(raw_policy or "")
    local_exists = False
    if _need_local_image_inspect(is_latest, policy):
        try:
            local_exists = await runner.image_exists(image_name)
        except Exception as error:
            return DockerBuildResult(False, f"inspect docker image error: {error}")

    if _if_pull_image(local_exists, is_latest, policy):
        if is_latest:
            post_log("pull latest image")
        post_log(f"start pull image: {image_name}")
        try:
            output = await runner.pull_image(
                image_name,
                str(credential.get("user", "")),
                str(credential.get("password", "")),
            )
            if output.strip():
                post_log(output)
        except Exception as error:
            return DockerBuildResult(False, f"pull image {image_name} error: {error}")
    else:
        post_log(f"use local exist image: {image_name}")

    try:
        (Path(options.work_dir) / _LOCAL_DOCKER_BUILD_TEMP).mkdir(parents=True, exist_ok=True)
    except OSError as error:
        return DockerBuildResult(False, f"create docker tmp dir error: {error}")
    suffix = options.rand_suffix() if options.rand_suffix else "".join(
        random.choice(string.ascii_letters) for _ in range(8)
    )
    container_name = (
        f"dispatch-{build.get('buildId', '')}-{build.get('vmSeqId', '')}-{suffix}"
    )
    try:
        create_args = await build_docker_create_args(container_name, image, options)
    except Exception as error:
        return DockerBuildResult(False, f"parse docker options error: {error}")
    try:
        container_id = await runner.create_container(create_args)
    except Exception as error:
        return DockerBuildResult(False, f"create container {container_name} error: {error}")

    try:
        try:
            await runner.start_container(container_id)
        except Exception as error:
            return DockerBuildResult(False, f"start container {container_name} error: {error}")
        try:
            status_code = await runner.wait_container(container_id)
        except Exception as error:
            return DockerBuildResult(False, f"wait container {container_name} error: {error}")
        if status_code != 0:
            message = await _read_docker_log_file(options, _LONG_LOG_TAG)
            try:
                container_log = await runner.container_logs(container_id)
            except Exception:
                container_log = ""
            if not message:
                message = container_log
            elif container_log.strip():
                post_log("docker container log: " + container_log)
            if message == _LONG_LOG_TAG:
                message = ""
            return DockerBuildResult(
                False, f"container {container_name} exit code {status_code}: {message}"
            )
        return DockerBuildResult(True, "")
    finally:
        try:
            await runner.remove_container(container_id)
        except Exception:
            pass


async def _read_docker_log_file(options: DockerBuildOptions, tag: str) -> str:
    build = options.build_info
    log_file = (
        Path(options.work_dir)
        / _LOCAL_DOCKER_WORKSPACE
        / "logs"
        / str(build.get("buildId", ""))
        / str(build.get("vmSeqId", ""))
        / "docker.log"
    )
    try:
        content = await asyncio.to_thread(log_file.read_bytes)
    except FileNotFoundError:
        return ""
    except OSError as error:
        return f"read log file error {error}"
    return tag if len(content) > 1000 else content.decode("utf-8", errors="replace")

File: src/agent/agent-sdk/nodejs/src/dockerBuild.ts
/*
 * BK-CI Agent SDK - docker 构建编排
 *
 * 移植 Go agent 的 src/pkg/job/build_docker.go 与 docker_runtime.go。
 * 用 docker/podman 拉起容器执行 worker：拉镜像 → create → start → wait → 取日志 → 结果。
 *
 * 容器内约定挂载（对齐 Go docker_runtime.go）：
 *   - worker-agent.jar → /data/worker-agent.jar（只读）
 *   - jdk17 目录       → /usr/local/jre（只读），jdk8 → /usr/local/jre8（只读）
 *     （仅 jdk8 时挂到 /usr/local/jre）
 *   - init.sh          → /data/init.sh（只读），entrypoint /bin/sh -c /data/init.sh
 *   - workspace/data   → DockerDataDir，logs → /data/devops/logs
 *
 * JDK env（DEVOPS_AGENT_JDK_*_PATH）仅在 linux 注入容器；mac/win 需用户自行设置（同 Go）。
 * 返回 {success, message}，由调用方用 api.workerBuildFinish 上报。
 */

import * as fs from 'fs';
import * as path from 'path';
import { DockerRunner } from './dockercli';
import { DockerOptions, ImagePullPolicy, ThirdPartyBuildInfo } from './types';

/** 容器内 worker.jar 路径。 */
export const ContainerWorkerJar = '/data/worker-agent.jar';
/** 容器内 init.sh 路径（entrypoint）。 */
export const EntryPointCmd = '/data/init.sh';
/** 容器内 jdk17 挂载点。 */
export const TargetJreDir = '/usr/local/jre';
/** 容器内 jdk8 挂载点。 */
export const TargetJre8Dir = '/usr/local/jre8';
/** 容器内日志目录。 */
export const DockerLogDir = '/data/devops/logs';
/** 容器内 workspace 数据目录（对应 Go constant_out.go DockerDataDir）。 */
export const DockerDataDir = '/data/devops/workspace';
/** 不挂载标记（对应 Go DockerNoMount）。 */
export const DockerNoMount = '__NO_MOUNT__';

const LocalDockerWorkSpaceDirName = 'docker_workspace';
const LocalDockerBuildTmpDirName = 'docker_build_tmp';
const longLogTag = 'toolong';

export interface DockerBuildOptions {
  /** 构建任务信息，须含 dockerBuildInfo。 */
  buildInfo: ThirdPartyBuildInfo;
  /** agent 工作目录。 */
  workDir: string;
  /** JDK17 目录（挂载到容器 /usr/local/jre）。为空则仅挂 jdk8 到 /usr/local/jre。 */
  jdk17DirPath?: string;
  /** JDK8 目录。 */
  jdk8DirPath?: string;
  /** worker-agent.jar 的宿主机绝对路径。 */
  workerJarPath: string;
  /** docker init 脚本的宿主机绝对路径（挂载到 /data/init.sh）。 */
  dockerInitScriptPath: string;
  /** 后台网关地址（写入容器 devops_gateway）。 */
  gateway: string;
  /** 项目 id（写入容器 devops_project_id）。 */
  projectId: string;
  /** 额外注入容器的环境变量。 */
  extraEnv?: Record<string, string>;
  /** 目标平台（决定是否注入 JDK env），默认当前平台。 */
  platform?: NodeJS.Platform;
  /** 日志上报回调（用于把 pull/容器日志转发到后台）。 */
  postLog?: (message: string) => void;
  /** 随机名后缀生成器（便于测试），默认随机 8 位。 */
  randSuffix?: () => string;
}

export interface DockerBuildResult {
  success: boolean;
  message: string;
}

function ensureGateway(gateway: string): string {
  if (gateway.startsWith('http://') || gateway.startsWith('https://')) return gateway;
  return 'http://' + gateway;
}

function randString(n: number): string {
  const chars = 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ';
  let s = '';
  for (let i = 0; i < n; i++) s += chars[Math.floor(Math.random() * chars.length)];
  return s;
}

// ── 用户 docker options 解析（对应 job_docker.parseApiDockerOptions）──
function parseApiDockerOptions(o: DockerOptions): string[] {
  const args: string[] = [];
  for (const v of o.volumes ?? []) {
    if (v.trim() === '') continue;
    args.push('--volume', v.trim());
  }
  for (const m of o.mounts ?? []) {
    if (m.trim() === '') continue;
    args.push('--mount', m.trim());
  }
  if ((o.gpus ?? '').trim() !== '') {
    args.push('--gpus', (o.gpus as string).trim());
  }
  if (o.privileged) {
    args.push('--privileged');
  }
  for (const n of o.network ?? []) {
    if (n.trim() === '') continue;
    args.push('--network', n.trim());
  }
  if ((o.user ?? '').trim() !== '') {
    args.push('--user', (o.user as string).trim());
  }
  return args;
}

function buildUserDockerArgs(o: DockerOptions): string[] {
  const argv = parseApiDockerOptions(o);
  for (let i = 0; i < argv.length; i++) {
    switch (argv[i]) {
      case '--volume':
      case '--mount':
      case '--network':
      case '--user':
      case '--gpus':
        if (i + 1 >= argv.length || argv[i + 1].trim() === '') {
          throw new Error(`docker option ${argv[i]} requires a non-empty value`);
        }
        i++;
        break;
    }
  }
  return argv;
}

function hasCustomNetwork(o: DockerOptions): boolean {
  return (o.network ?? []).length > 0;
}

function needLocalImageInspect(isLatest: boolean, policy: string): boolean {
  switch (policy) {
    case ImagePullPolicy.Always:
      return false;
    case ImagePullPolicy.IfNotPresent:
      return true;
    default:
      return !isLatest;
  }
}

function ifPullImage(localExist: boolean, isLatest: boolean, policy: string): boolean {
  switch (policy) {
    case ImagePullPolicy.Always:
      return true;
    case ImagePullPolicy.IfNotPresent:
      return !localExist;
    default:
      if (isLatest) return true;
      return !localExist;
  }
}

function hasJdk17Dir(jdk17DirPath?: string): boolean {
  if (!jdk17DirPath) return false;
  try {
    return fs.statSync(jdk17DirPath).isDirectory();
  } catch {
    return false;
  }
}

// ── 容器环境变量（对应 docker_runtime.go parseContainerEnv）──
function parseContainerEnv(opts: DockerBuildOptions): string[] {
  const d = opts.buildInfo.dockerBuildInfo;
  const platform = opts.platform ?? process.platform;
  const vars: string[] = [];
  vars.push('devops_project_id=' + opts.projectId);
  vars.push('devops_agent_id=' + (d?.agentId ?? ''));
  vars.push('devops_agent_secret_key=' + (d?.secretKey ?? ''));
  vars.push('devops_gateway=' + ensureGateway(opts.gateway));
  vars.push('agent_build_env=DOCKER');
  // mac/win 无法使用 agent 自带 JDK，仅 linux 注入
  if (hasJdk17Dir(opts.jdk17DirPath) && platform === 'linux') {
    vars.push('DEVOPS_AGENT_JDK_8_PATH=' + TargetJre8Dir + '/bin/java');
    vars.push('DEVOPS_AGENT_JDK_17_PATH=' + TargetJreDir + '/bin/java');
  }
  if (opts.extraEnv) {
    for (const [k, v] of Object.entries(opts.extraEnv)) {
      vars.push(`${k}=${v}`);
    }
  }
  return vars;
}

// ── mount 参数（对应 docker_runtime.go parseContainerMountArgs）──
async function parseContainerMountArgs(opts: DockerBuildOptions): Promise<string[]> {
  const b = opts.buildInfo;
  const args: string[] = [];

  if (hasJdk17Dir(opts.jdk17DirPath)) {
    args.push(
      '--mount',
      `type=bind,source=${opts.jdk17DirPath},target=${TargetJreDir},readonly`,
      '--mount',
      `type=bind,source=${opts.jdk8DirPath ?? ''},target=${TargetJre8Dir},readonly`
    );
  } else if (opts.jdk8DirPath) {
    args.push('--mount', `type=bind,source=${opts.jdk8DirPath},target=${TargetJreDir},readonly`);
  }

  args.push(
    '--mount',
    `type=bind,source=${opts.workerJarPath},target=${ContainerWorkerJar},readonly`,
    '--mount',
    `type=bind,source=${opts.dockerInitScriptPath},target=${EntryPointCmd},readonly`
  );

  let dataDir = path.join(
    opts.workDir,
    LocalDockerWorkSpaceDirName,
    'data',
    b.pipelineId,
    b.vmSeqId
  );
  if (b.workspace) {
    dataDir = b.workspace;
  }
  if (dataDir !== DockerNoMount) {
    await fs.promises.mkdir(dataDir, { recursive: true });
    args.push('--mount', `type=bind,source=${dataDir},target=${DockerDataDir}`);
  }

  const logsDir = path.join(
    opts.workDir,
    LocalDockerWorkSpaceDirName,
    'logs',
    b.buildId,
    b.vmSeqId
  );
  await fs.promises.mkdir(logsDir, { recursive: true });
  args.push('--mount', `type=bind,source=${logsDir},target=${DockerLogDir}`);

  return args;
}

/** 组装 docker create 参数（对应 docker_runtime.go buildDockerCreateArgs）。 */
export async function buildDockerCreateArgs(
  containerName: string,
  image: string,
  opts: DockerBuildOptions
): Promise<string[]> {
  const d = opts.buildInfo.dockerBuildInfo;
  const options: DockerOptions = d?.options ?? {};
  const mountArgs = await parseContainerMountArgs(opts);
  const userArgs = buildUserDockerArgs(options);

  const args: string[] = ['--name', containerName];
  args.push(...userArgs);
  if (!hasCustomNetwork(options)) {
    args.push('--network', 'bridge');
  }
  for (const e of parseContainerEnv(opts)) {
    args.push('-e', e);
  }
  args.push(...mountArgs);
  args.push('--entrypoint', '/bin/sh', image, '-c', EntryPointCmd);
  return args;
}

/**
 * 执行 docker 构建。runner 由调用方创建（可注入日志回调），opts 提供任务与路径信息。
 * 返回结果由调用方用 api.workerBuildFinish 上报（成功 sleep 8s 由调用方处理）。
 */
export async function runDockerBuild(
  runner: DockerRunner,
  opts: DockerBuildOptions
): Promise<DockerBuildResult> {
  const b = opts.buildInfo;
  const d = b.dockerBuildInfo;
  const postLog = opts.postLog ?? (() => undefined);

  if (!d) {
    return { success: false, message: 'dockerBuildInfo is missing' };
  }
  if (d.credential && d.credential.errMsg) {
    return { success: false, message: `get docker cred error: ${d.credential.errMsg}` };
  }

  const imageName = (d.image ?? '').trim();
  const imageStr = imageName.replace(/^https?:\/\//, '');

  const imageStrSub = imageStr.split(':');
  let isLatest = false;
  if (imageStrSub.length === 2 && imageStrSub[1] === 'latest') {
    isLatest = true;
  } else if (imageStrSub.length === 1) {
    isLatest = true;
  }
  const policy = d.imagePullPolicy ?? '';

  let localExist = false;
  if (needLocalImageInspect(isLatest, policy)) {
    try {
      localExist = await runner.imageExists(imageName);
    } catch (e) {
      return { success: false, message: `inspect docker image error: ${errMsg(e)}` };
    }
  }

  if (ifPullImage(localExist, isLatest, policy)) {
    if (isLatest) postLog('pull latest image');
    postLog(`start pull image: ${imageName}`);
    try {
      const out = await runner.pullImage(
        imageName,
        d.credential?.user ?? '',
        d.credential?.password ?? ''
      );
      if (out.trim() !== '') postLog(out);
    } catch (e) {
      return { success: false, message: `pull image ${imageName} error: ${errMsg(e)}` };
    }
  } else {
    postLog('use local exist image: ' + imageName);
  }

  // 创建 docker 构建临时空间
  const tmpDir = path.join(opts.workDir, LocalDockerBuildTmpDirName);
  try {
    await fs.promises.mkdir(tmpDir, { recursive: true });
  } catch (e) {
    return { success: false, message: `create docker tmp dir error: ${errMsg(e)}` };
  }

  const rand = opts.randSuffix ? opts.randSuffix() : randString(8);
  const containerName = `dispatch-${b.buildId}-${b.vmSeqId}-${rand}`;

  let createArgs: string[];
  try {
    createArgs = await buildDockerCreateArgs(containerName, imageStr, opts);
  } catch (e) {
    return { success: false, message: `parse docker options error: ${errMsg(e)}` };
  }

  let containerId: string;
  try {
    containerId = await runner.createContainer(createArgs);
  } catch (e) {
    return { success: false, message: `create container ${containerName} error: ${errMsg(e)}` };
  }

  try {
    try {
      await runner.startContainer(containerId);
    } catch (e) {
      return { success: false, message: `start container ${containerName} error: ${errMsg(e)}` };
    }

    let statusCode: number;
    try {
      statusCode = await runner.waitContainer(containerId);
    } catch (e) {
      return { success: false, message: `wait container ${containerName} error: ${errMsg(e)}` };
    }

    if (statusCode !== 0) {
      let msg = await readDockerLogFile(opts, longLogTag);
      const containerLog = await runner.containerLogs(containerId).catch(() => '');
      if (msg === '') {
        msg = containerLog;
      } else if (containerLog.trim() !== '') {
        postLog('docker container log: ' + containerLog);
      }
      if (msg === longLogTag) msg = '';
      return {
        success: false,
        message: `container ${containerName} exit code ${statusCode}: ${msg}`,
      };
    }

    return { success: true, message: '' };
  } finally {
    // 清理容器（对应 Go defer RemoveContainer）
    await runner.removeContainer(containerId).catch(() => undefined);
  }
}

/** 读取挂载出的 docker.log；超过 1000 字节返回 longLogTag（对应 Go logFile）。 */
async function readDockerLogFile(opts: DockerBuildOptions, tag: string): Promise<string> {
  const b = opts.buildInfo;
  const logFile = path.join(
    opts.workDir,
    LocalDockerWorkSpaceDirName,
    'logs',
    b.buildId,
    b.vmSeqId,
    'docker.log'
  );
  try {
    const content = await fs.promises.readFile(logFile);
    if (content.length > 1000) return tag;
    return content.toString('utf-8');
  } catch (e) {
    if ((e as NodeJS.ErrnoException).code === 'ENOENT') return '';
    return `read log file error ${errMsg(e)}`;
  }
}

function errMsg(e: unknown): string {
  return e instanceof Error ? e.message : String(e);
}
